### PR TITLE
10979 handle event of home screen

### PR DESCRIPTION
**Pull request by lamld-1814**
### Issue Link :link:

https://edu-redmine.sun-asterisk.vn/issues/10979
### Goals :soccer:

- Click the Favourite button to go to Favourite screen
- Click the Timer button to go to Timer screen
- Click the Mixer button to go to Mixer screen
- Click the Search button, the audio is searched
- Click Play/Pause button, the audio is turned on / turned off"
## Implementation Details :construction:

https://www.youtube.com/watch?v=joLgncCovLU&feature=youtu.be

### CODE REVIEW

File: SleepSound/SleepSound/Scenes/Main/MainViewController.swift [hunk 1]
```diff
@@ -11,19 +11,24 @@ final class MainViewController: UIViewController {
...
     @IBOutlet private weak var playPauseAudioButton: UIButton!
+    @IBOutlet private weak var audioCollectionView: UICollectionView!
     
     // MARK: - Properties
     private var arrAudio = [Audio]()
+    private var filteredAudio = [Audio]()
+    private var searchingAudio = false
     private let audioRepository = AudioRepositoryImpl(api: APIService.share)
     private var arrAudioPlayer = [String: AVPlayer]()
+    private var timerRandom: Timer! = nil
```
Comment (dongsonbk58) [inline SleepSound/SleepSound/Scenes/Main/MainViewController.swift]: sao Timer! lại gán bằng nil nhề

File: SleepSound/SleepSound/Scenes/Main/MainViewController.swift [hunk 2]
```diff
@@ -11,19 +11,24 @@ final class MainViewController: UIViewController {
...
 
     // MARK: - Life Cycle
     override func viewDidLoad() {
         super.viewDidLoad()
         configView()
         fetchData()
+        audioSearchBar.delegate = self
```
Comment (dongsonbk58) [inline SleepSound/SleepSound/Scenes/Main/MainViewController.swift]: ném vô cconfigView()

File: SleepSound/SleepSound/Scenes/Main/MainViewController.swift [hunk 3]
```diff
@@ -11,19 +11,24 @@ final class MainViewController: UIViewController {
...
     @IBOutlet private weak var mixerButton: UIButton!
-    @IBOutlet private weak var audioCollectionView: UICollectionView!
+    @IBOutlet private weak var randomButton: UIButton!
     @IBOutlet private weak var playPauseAudioButton: UIButton!
+    @IBOutlet private weak var audioCollectionView: UICollectionView!
     
     // MARK: - Properties
     private var arrAudio = [Audio]()
+    private var filteredAudio = [Audio]()
+    private var searchingAudio = false
```
Comment (tienpx3010) [inline SleepSound/SleepSound/Scenes/Main/MainViewController.swift]:
```suggestion
    private var isSearchingAudio = false
```

File: SleepSound/SleepSound/Scenes/Main/MainViewController.swift [hunk 4]
```diff
@@ -88,6 +102,43 @@ final class MainViewController: UIViewController {
...
     }
     
+    func stopTimerRandom() {
+        if timerRandom != nil {
+            timerRandom?.invalidate()
+        }
+    }
+    
+    func startTimerRandom() {
+        let indexPathRow = Int(arc4random_uniform(UInt32(arrAudio.count)))
```
Comment (tienpx3010) [inline SleepSound/SleepSound/Scenes/Main/MainViewController.swift]:
```suggestion
        let indexPathRow = Int.random(in: 0 ... arrAudio.count)
```

File: SleepSound/SleepSound/Scenes/Main/MainViewController.swift [hunk 5]
```diff
@@ -88,6 +102,43 @@ final class MainViewController: UIViewController {
...
+        let audio = arrAudio[indexPathRow].name
+        arrAudioPlayer.updateValue(playerForName(indexPathRow), forKey: audio)
+        reloadPlayButton()
```
Comment (tienpx3010) [inline SleepSound/SleepSound/Scenes/Main/MainViewController.swift]:
```suggestion
        configPlayButton()
```

File: SleepSound/SleepSound/Scenes/Main/MainViewController.swift [hunk 6]
```diff
@@ -88,6 +102,43 @@ final class MainViewController: UIViewController {
...
+        configPlayButton()
+        audioCollectionView.reloadData()
+    }
+    
+    func playerForName(_ row: Int) -> AVPlayer {
+        var player = AVPlayer()
+        if row >= Constants.arrAudioNameList.count {
+            let path = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]
+            if let pathComponent = NSURL(fileURLWithPath: path).appendingPathComponent("\(arrAudio[row].name).mp3") {
+                if !FileManager.default.fileExists(atPath: pathComponent.path) {
```
Comment (dongsonbk58) [inline SleepSound/SleepSound/Scenes/Main/MainViewController.swift]: if let pathComponent = NSURL(fileURLWithPath: path).appendingPathComponent("\(arrAudio[row].name).mp3"),
                 !FileManager.default.fileExists(atPath: pathComponent.path) {

ỏ chuyển qua guard let

Comment (dongsonbk58) [reply SleepSound/SleepSound/Scenes/Main/MainViewController.swift]: hạn chê dùng quá nhiều if else

File: SleepSound/SleepSound/Scenes/Main/MainViewController.swift [hunk 7]
```diff
@@ -119,23 +170,23 @@ extension MainViewController: UICollectionViewDelegateFlowLayout {
...
         var isCheckRowSelected = false
         var isCheckLocalImage = false
         for item in arrAudioPlayer where item.key == audio {
             isCheckRowSelected = true
         }
         let image = UIImage(named: "\(audio).png")
-        if indexPath.row >= Constants.arrAudioNameList.count {
-            isCheckLocalImage = indexPath.row >= Constants.arrAudioNameList.count
+        if image != nil {
+            isCheckLocalImage = false
```
Comment (dongsonbk58) [inline SleepSound/SleepSound/Scenes/Main/MainViewController.swift]: isCheckLocalImage = image == nil

**Review by dongsonbk58 — APPROVED** [review]

**Review by tgdong2296 — APPROVED** [review]

**Review by tienpx3010 — APPROVED** [review]

**Pull request merged**